tests/lwip/tcp.c

```diff
@@ -0,0 +1,223 @@
+    int res;
+
+    msg_init_queue(server_msg_queue, SERVER_MSG_QUEUE_SIZE);
+    /* parse port */
+    server_addr.port = (uint16_t)atoi((char *)args);
+    if ((res = sock_tcp_listen(&server_queue, &server_addr, &server_sock, 1,
+                               0)) < 0) {
+        printf("Unable to open TCP server on port %" PRIu16 " (error code %d)\n",
+               server_addr.port, -res);
+        return NULL;
+    }
+    server_running = true;
+    printf("Success: started TCP server on port %" PRIu16 "\n",
+           server_addr.port);
+    while (1) {
+        char client_addr[IPV6_ADDR_MAX_STR_LEN];
+        sock_tcp_t *sock = NULL;
+        int res;
+        unsigned client_port;
+
+        if ((res = sock_tcp_accept(&server_queue, &sock, SOCK_NO_TIMEOUT)) < 0) {
+            puts("Error on TCP accept");
+            continue;
+        }
+        else {
+            sock_tcp_ep_t client;
+
+            sock_tcp_get_remote(sock, &client);
+            ipv6_addr_to_str(client_addr, (ipv6_addr_t *)&client.addr.ipv6,
+                             sizeof(client_addr));
+            client_port = client.port;
+            printf("TCP client [%s]:%u connected\n",
+                   client_addr, client_port);
+        }
+        /* we don't use timeouts so all errors should be related to a lost
+         * connection */
+        while ((res = sock_tcp_read(sock, sock_inbuf, sizeof(sock_inbuf),
+                                    SOCK_NO_TIMEOUT)) >= 0) {
+            printf("Received TCP data from client [%s]:%u:\n",
+                   client_addr, client_port);
+            if (res > 0) {
+                od_hex_dump(sock_inbuf, res, 0);
+            }
+            else {
+                puts("(nul)");
+            }
+        }
+        printf("TCP connection to [%s]:%u reset, starting to accept again\n",
+               client_addr, client_port);
+        sock_tcp_disconnect(sock);
+    }
+    return NULL;
+}
+
+static int tcp_connect(char *addr_str, char *port_str, char *local_port_str)
+{
+    sock_tcp_ep_t dst = SOCK_IPV6_EP_ANY;
+    uint16_t local_port = 0;
+
+    if (client_running) {
+        puts("Cient already connected");
+    }
+
+    /* parse destination address */
+    if (ipv6_addr_from_str((ipv6_addr_t *)&dst.addr.ipv6, addr_str) == NULL) {
+        puts("Error: unable to parse destination address");
+        return 1;
+    }
+    /* parse port */
+    dst.port = (uint16_t)atoi(port_str);
+    if (local_port_str != NULL) {
+        local_port = (uint16_t)atoi(port_str);
+    }
+    if (sock_tcp_connect(&client_sock, &dst, local_port, 0) < 0) {
+        puts("Error: unable to connect");
+        return 1;
+    }
+    client_running = true;
+    return 0;
+}
+
+static int tcp_disconnect(void)
+{
+    sock_tcp_disconnect(&client_sock);
+    client_running = false;
+    return 0;
+}
+
+static int tcp_send(char *data, unsigned int num, unsigned int delay)
+{
+    uint8_t byte_data[strlen(data) / 2];
```

Comment:
(and where did we forbid dynamic arrays, especially in shell context where we know the maximum length this doesn't seem necessary)